cmd/kubectl/kubectlcobra/applier.go

```diff
@@ -0,0 +1,200 @@
+// Copyright 2019 The Kubernetes Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+package kubectlcobra
+
+import (
+	"context"
+	"time"
+
+	"github.com/go-errors/errors"
+	"github.com/spf13/cobra"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/cli-runtime/pkg/genericclioptions"
+	"k8s.io/cli-runtime/pkg/resource"
+	"k8s.io/kubectl/pkg/cmd/apply"
+	"k8s.io/kubectl/pkg/cmd/util"
+	"k8s.io/kubectl/pkg/scheme"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/kustomize/kstatus/wait"
+)
+
+// newApplier returns a new Applier. It will set up the applyOptions and
+// statusOptions which are responsible for capturing any command line flags.
+// It currently requires IOStreams, but this is a legacy from when
+// the ApplyOptions were responsible for printing progress. This is now
+// handled by a separate printer with the KubectlPrinterAdapter bridging
+// between the two.
+func newApplier(factory util.Factory, ioStreams genericclioptions.IOStreams) *Applier {
+	return &Applier{
+		applyOptions:  apply.NewApplyOptions(ioStreams),
+		statusOptions: NewStatusOptions(),
+		factory:       factory,
+		ioStreams:     ioStreams,
+	}
+}
+
+// resolver defines the interface the applier needs to observe status for resources.
+type resolver interface {
+	WaitForStatusOfObjects(ctx context.Context, objects []wait.KubernetesObject) <-chan wait.Event
+}
+
+// Applier performs the step of applying a set of resources into a cluster,
+// conditionally waits for all of them to be fully reconciled and finally
+// performs prune to clean up any resources that has been deleted.
+type Applier struct {
+	factory   util.Factory
+	ioStreams genericclioptions.IOStreams
+
+	applyOptions  *apply.ApplyOptions
+	statusOptions *StatusOptions
+	resolver      resolver
+}
+
+// Initialize sets up the Applier for actually doing an apply against
+// a cluster. This involves validating command line inputs and configuring
+// clients for communicating with the cluster.
+func (a *Applier) Initialize(cmd *cobra.Command) error {
+	a.applyOptions.PreProcessorFn = PrependGroupingObject(a.applyOptions)
```

Comment:
Note sure if we want this here. It would mean that users would have to call this `Initialize()` first if they wanted to override the PreProcessorFn. Let's not block on this now (i.e. we can change it later if that's what we want).